src/jit.c

```diff
@@ -0,0 +1,1557 @@
+    for (idx = 0, ir = block->ir_head; idx < block->n_insn; idx++, ir = next) {
+        next = ir->next;
+        ((codegen_block_func_t) dispatch_table[ir->opcode])(state, rv, ir);
+    }
+}
+
+static void resolve_jumps(struct jit_state *state)
+{
+    for (int i = 0; i < state->n_jumps; i++) {
+        struct jump jump = state->jumps[i];
+        int target_loc;
+        if (jump.target_offset != 0)
+            target_loc = jump.target_offset;
+        else if (jump.target_pc == TARGET_PC_EXIT)
+            target_loc = state->exit_loc;
+#if defined(__x86_64__)
+        else if (jump.target_pc == TARGET_PC_RETPOLINE)
+            target_loc = state->retpoline_loc;
+#elif defined(__aarch64__)
+        else if (jump.target_pc == TARGET_PC_ENTER)
+            target_loc = state->entry_loc;
+#endif
+        else {
+            target_loc = jump.offset_loc + sizeof(uint32_t);
+            for (int i = 0; i < state->n_insn; i++) {
+                if (jump.target_pc == state->offset_map[i].pc) {
+                    target_loc = state->offset_map[i].offset;
+                    break;
+                }
+            }
+        }
+#if defined(__x86_64__)
+        /* Assumes jump offset is at end of instruction */
+        uint32_t rel = target_loc - (jump.offset_loc + sizeof(uint32_t));
+
+        uint8_t *offset_ptr = &state->buf[jump.offset_loc];
+        memcpy(offset_ptr, &rel, sizeof(uint32_t));
+#elif defined(__aarch64__)
+        int32_t rel = target_loc - jump.offset_loc;
+        update_branch_immediate(state, jump.offset_loc, rel);
+#endif
+    }
+}
+
+static void translate_chained_block(struct jit_state *state,
+                                    riscv_t *rv,
+                                    block_t *block,
+                                    set_t *set)
+{
+    if (set_has(set, block->pc_start))
+        return;
+
+    set_add(set, block->pc_start);
+    offset_map_insert(state, block->pc_start);
+    translate(state, rv, block);
+    rv_insn_t *ir = block->ir_tail;
+    if (ir->branch_untaken && !set_has(set, ir->pc + 4)) {
+        block_t *block1 = cache_get(rv->block_cache, ir->pc + 4);
+        if (block1 && block1->translatable)
+            translate_chained_block(state, rv, block1, set);
+    }
+    if (ir->branch_taken && !set_has(set, ir->pc + ir->imm)) {
+        block_t *block1 = cache_get(rv->block_cache, ir->pc + ir->imm);
+        if (block1 && block1->translatable)
+            translate_chained_block(state, rv, block1, set);
+    }
+}
+
+uint32_t jit_translate(riscv_t *rv, block_t *block)
+{
+    struct jit_state *state = rv->jit_state;
+    memset(state->offset_map, 0, MAX_INSNS * sizeof(struct offset_map));
+    memset(state->jumps, 0, MAX_INSNS * sizeof(struct jump));
+    state->n_insn = 0;
+    state->n_jumps = 0;
+    uint32_t entry_loc = state->offset;
+    set_t set;
+    set_reset(&set);
+    translate_chained_block(&(*state), rv, block, &set);
+
+    if (state->offset == state->size) {
```

Comment:
If target buffer is too small, does it make sense to continue translating? Shall we raise abort or exception?

Reply:
In the function `emit_bytes`, we have ` assert(state->offset <= state->size - len);` to detect the machine code size is overflow or not.